### ISSUE TITLE
Losing Messages after Timeout

### ISSUE THREAD

**Issue by jackt-moran** [Azure/iotedge#1613]
We intermittently experience gaps of time where messages are lost somewhere in transit between edgeHub and IoT Hub (at least we think it's somewhere between the two). 

This behavior only occurs on edge devices in areas that occasionally have bad connectivity. However, we have some devices in areas of poor connectivity on which we've never experienced this issue.

We produce 2 messages per minute (1 every 30 seconds), so it is easy to see in the backend when we are missing data. We also consistently produce messages at this frequency across all our devices (and have logs to verify that), so we have ruled out this being an application error of ours.

Around the time where data is missing, the edgeHub logs indicate that it is disconnected and timing out trying to complete some operation.

## Expected Behavior

edgeHub should retry messages that could not be sent successfully because of a timeout.

## Current Behavior

As far as we can tell, when the timeout error occurs the message that edgeHub was trying to send is not persisted or retried. Of course, we could be misinterpreting what is happening behind the scenes.

## Steps to Reproduce

Note, these are steps to reproduce in a staging environment. We don't know exactly why our connection is intermittent so the best we can do is simulate a poor connection. Also note that these steps only work for hosts that can install Wondershaper, such as Ubuntu 18.04.

1. Deploy an Edge module that (approximately) produces 1 message every 30 seconds, each message 160 KB in size (bigger would also work - I don't think edgeHub accepts messages larger than 256 KB).
2. Install Wondershaper.
3. Determine edgeHubs network interface name (for us it is br-8f3c224237d5).
4. Run the command `wondershaper <network name> 1024 1024`. This limits <network name>'s downlink and uplink to 128 KB/s (the 1024's are in Kb).
5. Continue throttling the bandwidth for 20-30 minutes, and then clear the throttle.
6. Once edgeHub has finished catching up observe that not all data arrived in IoT Hub.

## Context (Environment)
Ubuntu 18.04

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```
√ config.yaml is well-formed - OK
√ config.yaml has well-formed connection string - OK
√ container engine is installed and functional - OK
√ config.yaml has correct hostname - OK
√ config.yaml has correct URIs for daemon mgmt endpoint - OK
‼ latest security daemon - Warning
    Error while fetching latest versions of edge components: could not send HTTP request
√ host time is close to real time - OK
√ container time is close to host time - OK
√ DNS server - OK
‼ production readiness: certificates - Warning
    Device is using self-signed, automatically generated certs.
    Please see https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: certificates expiry - OK
√ production readiness: container engine - OK
√ production readiness: logs policy - OK

Connectivity checks
-------------------
√ host can connect to and perform TLS handshake with IoT Hub AMQP port - OK
√ host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - OK
√ host can connect to and perform TLS handshake with IoT Hub MQTT port - OK
√ container on the default network can connect to IoT Hub AMQP port - OK
√ container on the default network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the default network can connect to IoT Hub MQTT port - OK
√ container on the IoT Edge module network can connect to IoT Hub AMQP port - OK
√ container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the IoT Edge module network can connect to IoT Hub MQTT port - OK
√ Edge Hub can bind to ports on host - OK
```
</details>

### Device (Host) Operating System
Ubuntu 18.04

### Architecture
amd64

### Container Operating System
Linux containers

#### iotedged

1.0.8

#### Edge Agent

1.0.8

#### Edge Hub

1.0.8

#### Docker

```
Client:
 Version:           3.0.6
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        a63faebc
 Built:             Wed Jun 19 22:56:37 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          3.0.6
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       b07f53d
  Built:            Wed Jun 19 22:28:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

## Logs

The edgeHub logs seem to have the most relevant information, so to start I've only included those. Please let me know if additional logs would be helpful.

<details>
<summary>iotedged logs</summary>

```
<Paste here>
```
</details>

<details>
<summary>edge-agent logs</summary>

```
<Paste here>
```
</details>

<details>
<summary>edge-hub logs</summary>
We think the error associated with our issue is the timeout near the bottom of our edgeHub logs starting at line 651.
[edgeHub logs](https://gist.github.com/jackt-moran/5961db9a4d839c9b90f65ca47ce39fb2)

</details>

## Additional Information

We think that the combination of our larger message sizes and intermittent connectivity are a part of the cause of the issue. We never had this issue when using smaller message sizes (for example 20KB).

We are also aware that non-production certificates could be causing issues. That said, we have two other devices also without production certs yet and we have not experienced this issue on those devices.

**Comment by ancaantochi** [comment]
I think it's not because the error where it fails to send the message because timeout (line 651), the next line I see it retries to send the message. It should be able to send it when the connectivity recovers.  Can you please provide full logs, I want to see if there are other exceptions and what are the steps when it recovers.

How to you determine that a message was lost, do you have the tracking id in the message?

**Comment by jackt-moran** [comment]
@ancaantochi,
We are using timestamps and device id together as unique identifiers. With those two pieces of info we build the path and file name of messages in our data lake. Because the module that produces our data publishes data at a rate of 1 message/30 seconds it is easy to see when there are missing messages. We also backup our data in a local PostgreSQL instance before sending to edgeHub, so we can verify the missing messages do exist and that we got verification from edgeHub via the C# SDK.

I am working on getting the full logs now I will post them when completed.

**Comment by jackt-moran** [comment]
@ancaantochi,

Below are links to more comprehensive logs containing multiple occurrences of the issue. Please let me know if you need any further information, thank you.

[iotedged](https://gist.github.com/jackt-moran/61cd78a45e9a7ad8e7d3b5c1365a7d6a)

[edgeAgent](https://gist.github.com/jackt-moran/2f535dd5cb873fc8d37de1fb03f3544c)

[edgeHub](https://gist.github.com/jackt-moran/9162dd36a6d09c7cfa79729811cb781a)

**Comment by veyalla** [comment]
@jackt-moran - we'd like to get more information about this and have private bits we'd like to share that may potentially help. Can you please reach out to me at veyalla@microsoft.com?

**Comment by veyalla** [comment]
@jackt-moran Closing this based on our offline conversation.

Note, we just released [1.0.8.3](https://github.com/Azure/azure-iotedge/releases/tag/1.0.8.3) which fixes a corner case condition where message loss may occur. Please use edge hub 1.0.8.3 as it is the most stable version with least known issues yet.
